external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala

```diff
@@ -404,20 +420,85 @@ abstract class KafkaSinkBatchSuiteBase extends KafkaSinkSuiteBase {
       .save()
     checkAnswer(
       createKafkaReader(topic, includeHeaders = true).selectExpr(
-        "CAST(value as STRING) value", "headers"
+        "CAST(value as STRING) value", "headers", "partition"
       ),
-      Row("1", Seq(Row("a", "b".getBytes(UTF_8)))) ::
-        Row("2", Seq(Row("c", "d".getBytes(UTF_8)), Row("e", "f".getBytes(UTF_8)))) ::
-        Row("3", Seq(Row("g", "h".getBytes(UTF_8)), Row("g", "i".getBytes(UTF_8)))) ::
-        Row("4", null) ::
+      Row("1", Seq(Row("a", "b".getBytes(UTF_8))), 0) ::
+        Row("2", Seq(Row("c", "d".getBytes(UTF_8)), Row("e", "f".getBytes(UTF_8))), 1) ::
+        Row("3", Seq(Row("g", "h".getBytes(UTF_8)), Row("g", "i".getBytes(UTF_8))), 2) ::
+        Row("4", null, 3) ::
         Row("5", Seq(
           Row("j", "k".getBytes(UTF_8)),
           Row("j", "l".getBytes(UTF_8)),
-          Row("m", "n".getBytes(UTF_8)))) ::
+          Row("m", "n".getBytes(UTF_8))), 0) ::
         Nil
     )
   }
 
+  test("batch - partition column vs default Kafka partitioner") {
```

Comment:
Given we add this test to check the functionality to specify partition, I think we don't need to change above test.

Reply:
I did it because a previous PR (headers) changed it as well. But I can revert it to the previous state if so desired.

Reply:
Personally yes, but we can hear more voices on this as well.

Reply:
I've gone through the test but not sure if it covers all. Here are my thought:
* Not sure why 100 partitions necessary?
* Don't we need the following tests? 1. `kafka.partitioner.class` overrides default 2. column overrides `kafka.partitioner.class` and default
* I think this test can be formed more simple. Put fixed key into a topic, get the partition (just like it is now but see my next comment). In another topic put all the data into a partition which is not the formerly get partition. Finally read back and double check.
* `.collect().toList.head` maybe not enough because it would be good to make sure the data is in one partition.
* There are couple of copy-pastes here as well.

Reply:
- "Not sure why 100 partitions necessary?": I just wanted to avoid the test to succeed by chance. I admit it's overkill. I'll reduce it to 10.
- "Don't we need the following tests?". See my other comment.
- "`.collect().toList.head` maybe not enough because it would be good to make sure the data is in one partition." `.collect()` docstring says "Returns an array that contains all rows in this Dataset. Running collect requires moving all the data into the application's driver process, and doing so on a very large dataset can crash the driver process with OutOfMemoryError." Does `collect()` work on a per-partition base? If so maybe I could use `.coalesce(1)` before calling `collect()`.
- "I think this test can be formed more simple." Ok it makes sense. I'll simplify it.
- "There are a couple of copy-pastes" Do you refer to the ```df.write
      .format("kafka")
      .option("kafka.bootstrap.servers", testUtils.brokerAddress)
      .option("topic", topic)
      .mode("append")
      .save()``` blocks? In that case, they are also all over the file.

Reply:
> I think `test("batch - write to kafka")` should be the simplest test and shouldn't contain neither headers nor partitions (I know for headers it's a bit late).

I can revert it to the previous state (with headers). Is that ok?

Reply:
> I can revert it to the previous state (with headers). Is that ok?

That's fine since that code is already in.

Reply:
> .collect() docstring says "Returns an array that contains all rows in this Dataset. Running collect requires moving all the data into the application's driver process, and doing so on a very large dataset can crash the driver process with OutOfMemoryError." Does collect() work on a per-partition base? If so maybe I could use .coalesce(1) before calling collect().

What I mean something like this:
`createKafkaReader(keyTopic).select("partition").map(_.getInt(0)).collect().toSet.size() === 1`

Reply:
> In that case, they are also all over the file.

Yes, the code can be enhanced several places but that doesn't mean we have to do the same copy-paste.

Reply:
> What I mean something like this:
> `createKafkaReader(keyTopic).select("partition").map(_.getInt(0)).collect().toSet.size() === 1`

I'm not sure I understand. Do you want me to assert that? Is this necessary considering that I send just one record?

Reply:
> Yes, the code can be enhanced several places but that doesn't mean we have to do the same copy-paste.

There are different opinions on copy-paste in UTs and I just copied the pattern in use in the file. But I also lean towards *not* copy-pasting. I can write a small function for that. Shall I apply it to other tests in the same function or just my contribution?

Reply:
In order not to have feature creep I would suggest to keep this for you change set. Later we can file another PR for the rest...